src/main/kotlin/org/kamiblue/client/module/modules/misc/NoteBot.kt

```diff
@@ -315,4 +320,88 @@ internal object NoteBot : Module(
             }
         }
     }
+
+    fun sortOutInstruments(){
+        if (!songName.value.endsWith(".nbs")) return
+
+        channelSettings[0].value = NoteBlockEvent.Instrument.PIANO
+        channelSettings[2].value = NoteBlockEvent.Instrument.BASSDRUM
+        channelSettings[3].value = NoteBlockEvent.Instrument.SNARE
+        channelSettings[4].value = NoteBlockEvent.Instrument.CLICKS
+        channelSettings[5].value = NoteBlockEvent.Instrument.GUITAR
+        channelSettings[6].value = NoteBlockEvent.Instrument.FLUTE
+        channelSettings[7].value = NoteBlockEvent.Instrument.BELL
+        channelSettings[8].value = NoteBlockEvent.Instrument.CHIME
+        channelSettings[9].value = NoteBlockEvent.Instrument.XYLOPHONE
+    }
+
+    fun readNbs(fileName: String): TreeMap<Long, ArrayList<Note>> {
+        val noteSequence = TreeMap<Long, ArrayList<Note>>()
+        val file = File(fileName)
+        val dataInputStream = DataInputStream(FileInputStream(file))
+        var length = dataInputStream.readShort()
+
+        var nbsversion = 0
+        if (length.toInt() == 0) {
+            nbsversion = dataInputStream.readByte().toInt()
+            dataInputStream.readByte().toInt()
+            if (nbsversion >= 3) {
+                length = dataInputStream.readShortCustom()
+            }
+        }
+        dataInputStream.readShortCustom()
+        dataInputStream.skipString()
+        dataInputStream.skipString()
+        dataInputStream.skipString()
+        dataInputStream.skipString()
+        val tempo = dataInputStream.readShortCustom()
+        val timeBetween = 1000 / (tempo / 100).toLong()
+        dataInputStream.skipBytes(23)
+        dataInputStream.skipString()
+        if (nbsversion >= 4) {
+            dataInputStream.skipBytes(4)
+        }
+        var currentTick: Short = -1
+        while (true) {
+            val jump = dataInputStream.readShortCustom()
+            if (jump.toInt() == 0) break
+            currentTick = (currentTick + jump).toShort()
+            var layer: Short = -1
+            while (true) {
+                val jumpLayer: Short = dataInputStream.readShortCustom()
+                if (jumpLayer == 0.toShort()) break
+                layer = (layer + jumpLayer).toShort()
+                val instrument = dataInputStream.readByte()
+                val key = dataInputStream.readByte()
+                if (nbsversion >= 4) {
+                    dataInputStream.readByte() // note block velocity
+                    dataInputStream.readByte() // note block panning
+                    dataInputStream.readShortCustom() // note block pitch
+                }
+                val time = timeBetween * currentTick
+                val note = key % 33
```

Comment:
Explain where this magical number is from

Reply:
https://opennbs.org/nbs from here, about half way down. The numbers are from 33 - 57 (the two octave range), however they could be from 0 - 87, so this just makes sure that they are in that range. It might be wrong, however listening to songs I can't hear it.

Reply:
Leave a comment explaining it then